what4/src/What4/Protocol/SMTLib2.hs

```diff
@@ -884,6 +919,344 @@ parseBvArraySolverValue w v (SApp ["store", arr, idx, val]) = do
+    case testEquality (I.exprType arg1_expr) (I.exprType arg2_expr) of
+      Just Refl -> liftIO (Some <$> I.isEq sym arg1_expr arg2_expr)
+      Nothing -> throwError ""
+  SApp ["ite", arg1, arg2, arg3] -> do
+    Some arg1_expr <- parseExpr sym arg1
+    Some arg2_expr <- parseExpr sym arg2
+    Some arg3_expr <- parseExpr sym arg3
+    case I.exprType arg1_expr of
+      I.BaseBoolRepr -> case testEquality (I.exprType arg2_expr) (I.exprType arg3_expr) of
+        Just Refl -> liftIO (Some <$> I.baseTypeIte sym arg1_expr arg2_expr arg3_expr)
+        Nothing -> throwError ""
+      _ -> throwError ""
+  SApp ["concat", arg1, arg2] -> do
+    Some arg1_expr <- parseExpr sym arg1
+    Some arg2_expr <- parseExpr sym arg2
+    BVProof{} <- getBVProof arg1_expr
+    BVProof{} <- getBVProof arg2_expr
+    liftIO $ Some <$> I.bvConcat sym arg1_expr arg2_expr
+--  SApp ["_", "extract", arg1, arg2, arg3] -> do
+
+  SApp ((SAtom operator) : operands) -> case HashMap.lookup operator (opTable @sym) of
+    Just (Op1 arg_types fn) -> do
+      args <- mapM (parseExpr sym) operands
+      exprAssignment arg_types args >>= \case
+        Ctx.Empty Ctx.:> arg1 ->
+          liftIO (Some <$> fn sym arg1)
+    Just (Op2 arg_types _ fn) -> do
+      args <- mapM (parseExpr sym) operands
+      exprAssignment arg_types args >>= \case
+        Ctx.Empty Ctx.:> arg1 Ctx.:> arg2 ->
+            liftIO (Some <$> fn sym arg1 arg2)
+    Just (BVOp1 op) -> do
+      Some arg_expr <- readOneArg sym operands
+      BVProof{} <- getBVProof arg_expr
+      liftIO $ Some <$> op sym arg_expr
+    Just (BVOp2 _ op) -> do
+      (Some arg1, Some arg2) <- readTwoArgs sym operands
+      BVProof m <- prefixError "in arg 1: " $ getBVProof arg1
+      BVProof n <- prefixError "in arg 2: " $ getBVProof arg2
+      case testEquality m n of
+        Just Refl -> liftIO (Some <$> op sym arg1 arg2)
+        Nothing -> throwError $ printf "arguments to %s must be the same length, \
+                                       \but arg 1 has length %s \
+                                       \and arg 2 has length %s"
+                                       operator
+                                       (show m)
+                                       (show n)
+    Just (BVComp2 op) -> do
+      (Some arg1, Some arg2) <- readTwoArgs sym operands
+      BVProof m <- prefixError "in arg 1: " $ getBVProof arg1
+      BVProof n <- prefixError "in arg 2: " $ getBVProof arg2
+      case testEquality m n of
+        Just Refl -> liftIO (Some <$> op sym arg1 arg2)
+        Nothing -> throwError $ printf "arguments to %s must be the same length, \
+                                       \but arg 1 has length %s \
+                                       \and arg 2 has length %s"
+                                       operator
+                                       (show m)
+                                       (show n)
+    -- Nothing -> throwError ""
+    _ -> throwError ""
```

Comment:
Is it possible to throw a more descriptive error here / at other instances of `throwError ""`? Also, I'd prefer to explicitly match `Nothing` vs. `_` if we're handling all cases.

Reply:
the plan is to merge back this code in the `What4.Serialize.Parser`, so I won't write error messages now.